File: src/main/java/org/web3j/gradle/plugin/Web3jExtension.java
/*
 * Copyright 2019 Web3 Labs Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
 * the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
 * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations under the License.
 */
package org.web3j.gradle.plugin;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

import org.gradle.api.Project;

import org.web3j.abi.datatypes.Address;

/** web3j extension for plugin configuration. */
public class Web3jExtension {

    /** Extension name used in Gradle build files. */
    public static final String NAME = "web3j";

    private static final String DEFAULT_GENERATED_PACKAGE = "org.web3j.{0}";

    /**
     * Generated package name for web3j contract wrappers. Accepts a {@link java.text.MessageFormat}
     * string with a unique parameter (i.e. {0} ), formatted as the contract name in lower case.
     */
    private String generatedPackageName;

    /** Base directory for generated Java files. */
    private String generatedFilesBaseDir;

    /** Generate smart contract wrappers using native Java types. */
    private Boolean useNativeJavaTypes;

    private Boolean generateBoth;

    /** Excluded contract names from wrapper generation. */
    private List<String> excludedContracts;

    /** Included contract names from wrapper generation. */
    private List<String> includedContracts;

    /** Bit length for network addresses. */
    private int addressBitLength;

    public String getGeneratedPackageName() {
        return generatedPackageName;
    }

    public void setGeneratedPackageName(final String generatedPackageName) {
        Objects.requireNonNull(generatedPackageName);
        this.generatedPackageName = generatedPackageName;
    }

    public String getGeneratedFilesBaseDir() {
        return generatedFilesBaseDir;
    }

    public void setGeneratedFilesBaseDir(final String generatedFilesBaseDir) {
        Objects.requireNonNull(generatedFilesBaseDir);
        this.generatedFilesBaseDir = generatedFilesBaseDir;
    }

    public Boolean getUseNativeJavaTypes() {
        return useNativeJavaTypes;
    }

    public void setUseNativeJavaTypes(final Boolean useNativeJavaTypes) {
        this.useNativeJavaTypes = useNativeJavaTypes;
    }

    public Boolean getGenerateBoth() {
        return generateBoth;
    }

    public void setGenerateBoth(Boolean generateBoth) {
        this.generateBoth = generateBoth;
    }

    public List<String> getExcludedContracts() {
        return excludedContracts;
    }

    public void setExcludedContracts(final List<String> excludedContracts) {
        this.excludedContracts = excludedContracts;
    }

    public List<String> getIncludedContracts() {
        return includedContracts;
    }

    public void setIncludedContracts(final List<String> includedContracts) {
        this.includedContracts = includedContracts;
    }

    public int getAddressBitLength() {
        return addressBitLength;
    }

    public void setAddressBitLength(final int addressBitLength) {
        this.addressBitLength = addressBitLength;
    }

    public Web3jExtension(final Project project) {
        generatedFilesBaseDir =
                project.getBuildDir().getAbsolutePath() + "/generated/sources/" + NAME;

        // Use the project's group name in generated package
        generatedPackageName = getDefaultGeneratedPackageName(project);

        useNativeJavaTypes = true;
        excludedContracts = new ArrayList<>();
        includedContracts = new ArrayList<>();
        addressBitLength = Address.DEFAULT_LENGTH / Byte.SIZE;
        generateBoth = false;
    }

    protected String getDefaultGeneratedPackageName(Project project) {
        String defaultPackageName;
        final String projectGroup = project.getGroup().toString();
        if (!projectGroup.isEmpty()) {
            defaultPackageName = projectGroup + "." + NAME;
        } else {
            defaultPackageName = DEFAULT_GENERATED_PACKAGE;
        }
        return defaultPackageName;
    }
}

File: src/main/kotlin/org/web3j/gradle/plugin/Web3jExtension.kt
/*
 * Copyright 2019 Web3 Labs Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
 * the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
 * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations under the License.
 */
package org.web3j.gradle.plugin

import org.gradle.api.Project
import org.web3j.abi.datatypes.Address

/** web3j extension for plugin configuration.  */
open class Web3jExtension(project: Project) {
    /**
     * Generated package name for web3j contract wrappers. Accepts a [java.text.MessageFormat]
     * string with a unique parameter (i.e. {0} ), formatted as the contract name in lower case.
     */
    open var generatedPackageName: String = getDefaultGeneratedPackageName(project)

    /** Base directory for generated Java files.  */
    var generatedFilesBaseDir = project.buildDir.absolutePath + "/generated/sources/" + NAME

    /** Generate smart contract wrappers using native Java types.  */
    var useNativeJavaTypes: Boolean = true

    /** Excluded contract names from wrapper generation.  */
    var excludedContracts: List<String> = emptyList()

    /** Included contract names from wrapper generation.  */
    var includedContracts: List<String> = emptyList()

    /** Bit length for network addresses.  */
    var addressBitLength: Int = Address.DEFAULT_LENGTH / java.lang.Byte.SIZE

    protected open fun getDefaultGeneratedPackageName(project: Project): String {
        val defaultPackageName: String
        val projectGroup = project.group.toString()
        defaultPackageName = if (projectGroup.isNotEmpty()) {
            "$projectGroup.$NAME"
        } else {
            DEFAULT_GENERATED_PACKAGE
        }
        return defaultPackageName
    }

    companion object {
        /** Extension name used in Gradle build files.  */
        const val NAME = "web3j"

        private const val DEFAULT_GENERATED_PACKAGE = "org.web3j.{0}"
    }
}
